### PR TITLE
fix: fail loud on concurrent-commit test corruption, guard commits against in-flight pushes

### PR DESCRIPTION
**Pull request by rjmurillo**
## Summary

Issue #5123 reported that committing to a worktree while its own background pre-push test run is still using it corrupts git-HEAD-reading fixtures and burns the full 11-minute suite. The issue asked for three things; this PR implements the first two and explicitly scopes out the third.

- **Ask #1 (detect structurally, fail loud).** `conftest.py`'s `_guard_real_repo_head` fixture already warned (`#3109`) when it detected a concurrent external commit during a test's window, but only warned: a test that failed because HEAD moved under it still showed a plain, misleading `AssertionError` as its primary failure. A new `pytest_runtest_makereport` hook stashes whether the test's call phase failed; the fixture now escalates to a distinct, greppable `#5123` `pytest.fail` when the failing test also saw an unattributed HEAD move, instead of only warning. A passing test that merely observed a concurrent HEAD change still just warns, since most tests never read git state and the result stays trustworthy.
- **Ask #2 (prevent the collision).** New `scripts/validation/check_push_lock_before_commit.py`, wired in as the `push-lock-commit-guard` pre-commit job, refuses a commit while a push for the current branch name is holding the canonical push-lock file (`.claude/rules/push-lock.md`). It reuses that exact lock file with a non-blocking probe rather than inventing a second locking scheme (push-lock.md MUST NOT 2). Git runs the pre-push hook inside the `git push` invocation the lock already wraps, so the lock stays held for the whole suite, not just the ref transfer.
- **Ask #3 (isolate git-HEAD-reading fixtures behind a throwaway repo)**: out of scope for this PR. It is a broader test-suite refactor (multiple test files, e.g. `tests/skills/memory/test_repair_episode_causal_links.py`), not a bounded fix. Recommending it as a follow-up rather than implementing it here.

## Specification References

| Type | Reference | Description |
|------|-----------|-------------|
| **Issue** | Refs #5123 | Committing to a worktree while its own background pre-push test run is still using it corrupts git-HEAD-reading fixtures and burns the full 11-minute suite |

## Changes

- `conftest.py`, `tests/test_pytest_head_guard.py`: escalate the concurrent-HEAD-change detector to a distinct `#5123` failure when the test itself already failed, instead of only warning.
- `scripts/validation/check_push_lock_before_commit.py`, `tests/validation/test_check_push_lock_before_commit.py`, `lefthook.yml`, `.claude/rules/push-lock.md`: new pre-commit guard that refuses a commit while this branch's own push is in flight, reusing the canonical push-lock file.
- Follow-up fixes from three rounds of automated review (security agent, Copilot, Cursor Bugbot, Devin): `SKIP_PUSH_LOCK_COMMIT_GUARD=1` escape hatch, fail-open on git errors, corrected "in this worktree" to the guard's real scope ("per branch name per machine"), a dead-code fixture-injection bug (`request` parameter defaulted to `None`, silently disabling the `#5123` escalation in production), Windows test coverage for the guard's `msvcrt.locking` branch, an overclaimed causal statement in the escalation message, and a config-catalog citation kept current against docstring line shifts.
- `.agents/retrospective/2026-08-25-issue-5123-push-lock-and-head-guard.md`: session retrospective recording the review catches, including the dead-code bug as a third instance of the same failure pattern.

## Security review

`lefthook.yml` changes route through the repo's `infrastructure-advisory` hook to a mandatory security-agent review (security.md MUST-1) before push. That review returned **APPROVE-WITH-FIXES**: no exploitable vulnerability (command injection, path traversal, symlink/arbitrary-write risk all checked with concrete repro attempts, all clean), but it found a real availability defect. With no bypass and fail-closed behavior on any git error, an orphaned lock holder (a killed pre-push job leaving a background child running, per lefthook's `timeout:` kills) could block every commit on that branch indefinitely. It also caught a false "in this worktree" scoping claim: the canonical lock path carries no worktree component by design (push-lock.md MUST 1), so the guard's real scope is per branch name per machine.

Both are fixed: `SKIP_PUSH_LOCK_COMMIT_GUARD=1` escape hatch (matching the existing `SKIP_YAMLLINT`/`FORCE_PUSH_OK` convention), fail-open on git errors with stderr surfaced, and corrected wording throughout. Demonstrated live against this repo: holding the real canonical lock for this branch made the guard refuse a commit with exit 1 and the accurate "on this machine" message; releasing it returned exit 0.

Per security.md MUST-5, no threat-model benchmark update was made. Neither CWE-22 (path traversal: `git check-ref-format --branch` rejects every traversal-shaped branch name before it reaches the lock-path string) nor CWE-77/78 (command injection: every subprocess call passes a list argv with no shell) is reachable here. This is confirmed by both the security review's manual repro attempts and a clean local `security-scan` run.

## Test plan

- [x] `tests/test_pytest_head_guard.py`: positive/negative/edge cases for the `call_failed` escalation (escalate-and-fail, stay-a-warning on pass, attributed-mutation ignores `call_failed`, hook wiring both directions) plus a real-pytest-subprocess end-to-end probe that exercises actual fixture injection rather than bypassing it.
- [x] `tests/validation/test_check_push_lock_before_commit.py`: positive/negative/edge cases for the lock probe on both the POSIX `fcntl.flock` and Windows `msvcrt.locking` branches, the branch-scoped path, the fail-open paths, the bypass env var, and the CLI exit-code contract. Exact counts are not quoted here since they drifted stale across review rounds; run `--collect-only -q` on both files for the current figure.
- [x] Full local pre-push suite green, including `python-tests` (28,000+ items): this is the exact suite issue #5123 describes as expensive to burn on a false failure.
- [x] Live repro against the real repo: holding the actual canonical push-lock file for this branch blocked a commit with the corrected message; releasing it allowed the commit.
- [x] `ruff`, `mypy`, `lefthook validate`, `check_push_lock_paths.py`, `check_skill_md_portability.py` all clean.

## Retrospective

`.agents/retrospective/2026-08-25-issue-5123-push-lock-and-head-guard.md` records the review catches as Confident-Incorrectness Recurrences (FAILURE-MODES.md #9), including a dead-code fixture-injection bug that shipped past the security review entirely and was caught only by later automated code review.

Refs #5123

### CODE REVIEW

File: conftest.py [hunk 1]
```diff
@@ -439,18 +475,33 @@ def _check_head_change(
...
+            "to finish, then re-run.",
+            pytrace=False,
+        )
+    warnings.warn(f"#3109: {concurrent_commit_detail}", stacklevel=2)
 
 
 @pytest.fixture(autouse=True)
-def _guard_real_repo_head() -> Iterator[None]:
-    """Attribute project HEAD movement without blaming concurrent commits."""
+def _guard_real_repo_head(request: pytest.FixtureRequest | None = None) -> Iterator[None]:
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply conftest.py]: Confirmed and fixed in `35a773096`. You were right: pytest's `getfuncargnames` excludes any parameter with a default from fixture injection, so `request` stayed `None` on every real run and the whole `#5123` escalation was dead code. Removed the default, updated the six direct `__wrapped__()` call sites in `tests/test_pytest_head_guard.py` to pass a fake request, and added `test_real_pytest_run_escalates_when_a_concurrent_commit_fails_the_test`, a real nested-pytest-subprocess integration test that exercises actual fixture resolution (not `__wrapped__()`). Verified it discriminates: reintroducing the default makes it fail with `AttributeError: 'NoneType' object has no attribute 'node'`, the exact symptom of the bug.

Recorded as a third instance of the same failure pattern in the retrospective (`.agents/retrospective/2026-08-25-issue-5123-push-lock-and-head-guard.md`).

---
_Generated by [Claude Code](https://claude.ai/code)_

[the remark this answers is not in the document]
Comment (rjmurillo) [reply conftest.py]: Same finding as the other reviewer's comment on this line, already fixed in `35a773096` (which landed before this review posted, though the push lagged). `request` no longer carries a default; the six direct `__wrapped__()` call sites in `tests/test_pytest_head_guard.py` were updated to pass a fake request, and a new subprocess-based integration test drives real pytest fixture resolution end to end. See the sibling reply on `conftest.py` line 497 for the verification detail.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: lefthook.yml [hunk 2]
```diff
@@ -30,6 +30,16 @@ pre-commit:
...
 
+    # Issue #5123: a commit landing while a push for this branch name is
+    # still running its pre-push suite (6-15 minutes), in this worktree or
+    # any other checkout of the same branch on this machine, can corrupt any
+    # test whose fixture reads live git state. Reuses the existing per-branch
+    # push-lock file (push-lock.md) rather than a second scheme; blocks only
+    # when that lock is currently held. Bypass: SKIP_PUSH_LOCK_COMMIT_GUARD=1.
+    - name: push-lock-commit-guard
+      timeout: 30s
+      run: uv run --frozen python scripts/validation/check_push_lock_before_commit.py
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply lefthook.yml:41]: Agreed, and this is a real limitation, not a misunderstanding: the probe-and-release design narrows the race, it does not close it. Documented explicitly rather than silently accepted, in two places:

- `scripts/validation/check_push_lock_before_commit.py` docstring, "Known limitation" paragraph: states the guard runs early in the piped sequence, releases the lock immediately, and a push can still acquire it during a later pre-commit job before `git commit` finishes.
- `.claude/rules/push-lock.md`, new "Commit guard" section, matching "Known limitation" paragraph, naming what full closure would require (commit and push sharing one held lock across their complete mutation windows).

Closing it fully is a larger design change than this PR's scope (issue #5123 asked for prevention of the collision, and this narrows it substantially); flagging that as a legitimate follow-up rather than doing it here.

---
_Generated by [Claude Code](https://claude.ai/code)_

[the remark this answers is not in the document]
Comment (rjmurillo) [reply lefthook.yml:41]: Fixed in `452b41ca9`. Added `push-lock-commit-guard` to the `expected_pre_commit` set in `test_configuration_uses_named_native_jobs()` plus explicit assertions on its `run` path, `timeout`, and its ordering before `markdown-autofix`. Verified it discriminates with a temporary lefthook.yml job-removal negative control before landing it.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/validation/check_push_lock_before_commit.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+where ``<slug>`` is the current branch name with every ``/`` replaced by
+``-``. This checker only ever reads that path's lock state; it never creates
+or holds the lock itself beyond the instant it takes to probe it.
+
+A refused commit is not lost: wait for the in-flight push to finish (or its
+pre-push suite to fail and release the lock), then commit again. If the lock
+is stuck (the holder crashed without releasing it, or you need to commit
+regardless), set ``SKIP_PUSH_LOCK_COMMIT_GUARD=1`` to bypass this one check,
+matching the ``SKIP_YAMLLINT``/``SKIP_CLI_E2E``/``FORCE_PUSH_OK`` escape-hatch
+convention in ``scripts/validation/git_hook_policy.py``.
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/validation/check_push_lock_before_commit.py]: Correct, and thank you for catching it: I described the behavior from memory instead of reading `git_hook_policy.py` first. Fixed in `342992121`. The docstring now cites the exact lines and states plainly that `SKIP_CLI_E2E=true` is rejected (exit 2, "cannot bypass a required CLI E2E gate"), not a bypass precedent.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: .agents/retrospective/2026-08-25-issue-5123-push-lock-and-head-guard.md [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+  the fail-open paths, and the bypass env var) and
+  `tests/test_pytest_head_guard.py` additions for the `call_failed`
+  escalation (9 new cases: escalate-and-fail, stay-a-warning,
+  attributed-mutation-ignores-call_failed, hook wiring both directions).
+
+## Remediation / follow-ups
+
+- No governance or instruction change proposed. The existing
+  `infrastructure-advisory` routing to security review already caught this;
+  the gap was in my own pre-review diligence, not in the repo's process.
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply .agents/retrospective/2026-08-25-issue-5123-push-lock-and-head-guard.md]: Fixed in `d9329971a`, which records the third Confident-Incorrectness instance (the dead-default-parameter bug both this PR's own security review and the two automated reviews on this thread caught) as remediation, while keeping the claim accurate: no separate governance-process change is proposed, only the instruction change this PR itself makes (`push-lock.md`'s new "Commit guard" section and its generated mirror).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/validation/check_push_lock_before_commit.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+from __future__ import annotations
+
+import argparse
+import os
+import subprocess
+import sys
+from collections.abc import Sequence
+from pathlib import Path
+
+SKIP_ENV = "SKIP_PUSH_LOCK_COMMIT_GUARD"
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/validation/check_push_lock_before_commit.py:71]: Fixed in `342992121`: added the `SKIP_PUSH_LOCK_COMMIT_GUARD` row to `.claude/skills/ai-agents-config-catalog/SKILL.md`'s "Environment Variables: Tests and Tooling" table, a provenance re-verify one-liner, and the missing bad-value edge-case test (`test_main_does_not_bypass_on_a_non_1_value`, parametrized over `["0", "true", "yes", ""]`).

Two follow-up commits (`0f42efa97`, `2bcb83d27`) were needed to regenerate the stale `src/copilot-cli/` mirror and raise the vendor-portability baseline for the new reference, both caught by this repo's own pre-push gates.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/validation/test_check_push_lock_before_commit.py [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+# of this whole file on Windows CI (ModuleNotFoundError), taking down every
+# platform-agnostic test in it too, so the import itself is guarded here.
+if sys.platform != "win32":
+    import fcntl
+
+_posix_only = pytest.mark.skipif(
+    sys.platform == "win32",
+    reason="exercises the POSIX fcntl.flock branch of _push_is_in_flight; "
+    "the Windows msvcrt branch has no equivalent test here",
+)
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply tests/validation/test_check_push_lock_before_commit.py:48]: Fixed in `e7cc5dd62`. Added three Windows-marked tests mirroring the existing POSIX pattern (a second same-process `open()` contends for the same `msvcrt` lock, matching this file's existing `fcntl.flock` convention): lock-held, lock-released, and `main()` exits 1 while held. Marked `pytest.mark.windows_path` so `.github/workflows/pytest.yml:603-607`'s `pytest -m windows_path` job on `windows-latest` actually executes them; `windows_path` is additive, not exclusive (no `-m "not windows_path"` deselection on the main job), so they also run in the normal Linux suite, skipped there with a named reason rather than silently absent.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: .claude/skills/ai-agents-config-catalog/SKILL.md [hunk 7]
```diff
@@ -61,6 +61,7 @@ Lesson encoded: a global bypass with no teeth (no telemetry, no approval step) w
...
 | pytest markers `unit`, `integration`, `safe_push_transport`, `security`, `smoke`, `windows_path` | pytest -m selectors | Filter test classes; `smoke` = real-CLI tests needing auth/credits, nightly only, and the smoke gate asserts they were NOT skipped (issue #2231 item 4); `safe_push_transport` = touches a non-local transport, excluded from pre-push | Production | Marking a test `smoke` to dodge CI is detected by the not-skipped assertion | `pyproject.toml [tool.pytest.ini_options].markers` |
+| `SKIP_PUSH_LOCK_COMMIT_GUARD` | env var (`=1` exact match) | Bypasses the `push-lock-commit-guard` pre-commit job (issue #5123), which otherwise refuses a commit while a push for the current branch name holds the canonical push-lock file | Production escape hatch | Scoped to one check, not global. Prints `push-lock: commit guard bypassed via SKIP_PUSH_LOCK_COMMIT_GUARD=1` so the bypass is visible in commit output; any other value (`0`, `true`, unset) leaves the guard enforced | `scripts/validation/check_push_lock_before_commit.py:58,165-166`; wired as `push-lock-commit-guard` in `lefthook.yml` |
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply .claude/skills/ai-agents-config-catalog/SKILL.md]: Fixed in `3a0362c77`. Corrected the citation to `69,176-178` (the actual `SKIP_ENV` definition and activation) and regenerated the `src/copilot-cli/` mirror in the same commit.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: .agents/retrospective/2026-08-25-issue-5123-push-lock-and-head-guard.md [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+
+## What shipped
+
+Issue #5123 reported that a commit landing in a worktree while that worktree's
+own pre-push suite is still running can corrupt any test whose fixture reads
+live git state, because the concurrent-commit detector only warned
+(`#3109`) instead of failing loud, and nothing prevented the collision in the
+first place.
+
+Three commits on `claude/autoplan-goal-joliu2`:
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply .agents/retrospective/2026-08-25-issue-5123-push-lock-and-head-guard.md:17]: Fixed in `3a0362c77`. Added a follow-up-commits summary right after the original three, distinguishing them from each later review-fix commit and grouping by what it addressed, so the retrospective reads as an accurate record rather than the initial submission's snapshot.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: conftest.py [hunk 9]
```diff
@@ -439,18 +475,39 @@ def _check_head_change(
...
-        stacklevel=2,
+        "so this is a concurrent external commit in this worktree."
     )
+    if call_failed:
+        pytest.fail(
+            f"#5123: {concurrent_commit_detail} The repository moved under "
+            "this test while it read live Git state, so the test's own "
+            "failure above is not meaningful. Do not debug it as a code "
+            "defect: wait for the in-flight push or commit in this worktree "
+            "to finish, then re-run.",
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply conftest.py]: Correct: temporal overlap is all this fixture can actually establish, and the old wording overclaimed it. Fixed in `81b92aaea`, reworded to say what's actually known ("that is correlation, not proof: this autouse fixture cannot tell whether this specific test actually read live Git state") and softened the conclusion from "is not meaningful" to "may not be meaningful," keeping the same re-run guidance. Updated both test assertions that pinned the old literal substring, verified the negative control (reintroducing the dead-default-parameter defect) still discriminates correctly against the new wording.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/validation/skill_md_portability_baseline.json [hunk 10]
```diff
@@ -300,7 +300,7 @@
     ".claude/skills/ai-agents-change-control/references/gate-ladder.md": 5,
     ".claude/skills/ai-agents-change-control/references/incident-history.md": 8,
     ".claude/skills/ai-agents-change-control/references/provenance.md": 14,
-    ".claude/skills/ai-agents-config-catalog/SKILL.md": 23,
+    ".claude/skills/ai-agents-config-catalog/SKILL.md": 24,
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/validation/skill_md_portability_baseline.json]: Confirmed and fixed properly rather than re-grandfathered, in `3155a2b16`. Added the concrete `scripts/validation/check_push_lock_before_commit.py` path to the canonical marker (`.claude/skills/ai-agents-config-catalog/SKILL.md`), regenerated the mirror, and re-ran the checker: drift dropped from 24 back to 23, an improvement over the original baseline rather than a new grandfather entry. Reverted both `drift_files` entries to 23 with `--allow-baseline-shrink`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: .claude/skills/ai-agents-config-catalog/SKILL.md [hunk 11]
```diff
@@ -61,6 +61,7 @@ Lesson encoded: a global bypass with no teeth (no telemetry, no approval step) w
...
 | pytest markers `unit`, `integration`, `safe_push_transport`, `security`, `smoke`, `windows_path` | pytest -m selectors | Filter test classes; `smoke` = real-CLI tests needing auth/credits, nightly only, and the smoke gate asserts they were NOT skipped (issue #2231 item 4); `safe_push_transport` = touches a non-local transport, excluded from pre-push | Production | Marking a test `smoke` to dodge CI is detected by the not-skipped assertion | `pyproject.toml [tool.pytest.ini_options].markers` |
+| `SKIP_PUSH_LOCK_COMMIT_GUARD` | env var (`=1` exact match) | Bypasses the `push-lock-commit-guard` pre-commit job (issue #5123), which otherwise refuses a commit while a push for the current branch name holds the canonical push-lock file | Production escape hatch | Scoped to one check, not global. Prints `push-lock: commit guard bypassed via SKIP_PUSH_LOCK_COMMIT_GUARD=1` so the bypass is visible in commit output; any other value (`0`, `true`, unset) leaves the guard enforced | `scripts/validation/check_push_lock_before_commit.py:69,176-178`; wired as `push-lock-commit-guard` in `lefthook.yml` |
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply .claude/skills/ai-agents-config-catalog/SKILL.md]: Fixed in `3155a2b16`: `71,177-180`. This citation drifted twice because an earlier commit in this same PR added lines to the module docstring above it, shifting everything below. Kept the existing line-number convention (matches every other row in this catalog's "Where defined" column) rather than changing format for one row.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/validation/check_push_lock_before_commit.py [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+        errors="replace",
+        check=True,
+        timeout=10,
+    )
+    return result.stdout.strip() or None
+
+
+def lock_path_for_branch(branch: str, lock_directory: Path | None = None) -> Path:
+    """Return the canonical push-lock path for ``branch`` (push-lock.md)."""
+    slug = branch.replace("/", "-")
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/validation/check_push_lock_before_commit.py:108]: Verified the collision is real, but it's not something this PR introduces: `.claude/rules/push-lock.md` MUST 2, which predates this PR, already defines the canonical scheme as bare `tr '/' '-'` (line 33's shell recipe), and this PR's own new "Commit guard" section explicitly documents the exact inherited collapse you're describing: "two branch names that only differ by `/` (`a/b` and `a-b`) share one lock file and therefore one guard" (push-lock.md, in the section this PR adds). I named it deliberately rather than missing it.

I can't fix it narrowly inside `check_push_lock_before_commit.py` alone: this guard's whole purpose is to probe the *same* lock file a real `git push` holds, so `lock_path_for_branch()` has to stay byte-identical to the canonical recipe every push in this repo uses (push-lock.md MUST NOT 2: "MUST NOT introduce a second lock path 'just for this run'"). A collision-free encoding change means rewriting the shell recipe quoted in that rule file and used verbatim in every push across the repo, not just this guard, which is exactly the kind of repo-wide, cross-cutting change push-lock.md's own design (issue #4366: "one path, written the same way everywhere") exists to prevent from proliferating.

I read `MUST 2`'s "two distinct branches must never share a lock" as motivated by the old hashed four-slot scheme's ~25% *probabilistic* collision rate (issue #4366), not as a guarantee this deterministic `/`-for-`-` collapse already breaks for the specific case of two branches that differ only by that one character. That's a real gap in the rule's own internal consistency, but closing it is a design decision for the canonical scheme itself, not something to bundle into a PR whose job is reusing that scheme as-is. Recommending a follow-up issue against `push-lock.md` to either accept and reword MUST 2, or pick and roll out a collision-free encoding repo-wide. Not implementing that here, the same way Ask #3 of issue #5123 was already scoped out of this PR.

---
_Generated by [Claude Code](https://claude.ai/code)_
